### DISCUSSION TITLE
[Script request] Calibre-Web Upgrade or Option for Calibre-Web-Automated

### DISCUSSION THREAD

**Discussion by Brixxor** [community-scripts/ProxmoxVE#125] [Request script]
### Application Name

Calibre-Web-Automated 

### Website

https://github.com/crocodilestick/Calibre-Web-Automated

### Description

I'm a big fan of the scripts. I know that the Calibre Web Script already exists, but the programme I am proposing is more than just an extension. The range of functions is many times greater and therefore my question is whether this would be more of an ‘extra’ script or whether there could be a selection in the installation process / deployment which of the two existing scripts should be installed?

### Due Diligence

- [X] I have searched existing [scripts](https://community-scripts.github.io/Proxmox/scripts) and found no duplicates.
- [X] I have searched existing [discussions](https://github.com/community-scripts/ProxmoxVE/discussions?discussions_q=) and found no duplicate requests.
- [X] This is not a game-related request.

**Comment by aibosss** [comment]
+1 for Calibre-Web-Automated  please please please

**Comment by xekil** [comment]
+2 I would like that too

**Comment by pladium1985** [comment]
+3

**Comment by Builder-DE-TH** [comment]
+4

**Comment by aibosss** [comment]
just to let you know guys I've used the portainer LXC and installed the docker CalibreWebAutomated in 2minutes

**Reply by pladium1985** [reply]
Will try, thanks!

**Comment by Creepo** [comment]
Maybe it would possible to implement this too into the LXC for this? https://github.com/calibrain/calibre-web-automated-book-downloader

**Comment by Tooelite** [comment]
I do not like to use Docker inside Proxmox. How about this:

- install a "calibre-web" lxc as base install, name it "calibre-web-automated"
- login into the proxmox lxc container, go to the home folder
- install git
- clone the repository https://github.com/crocodilestick/Calibre-Web-Automated to the home folder
- analyse the docker-file and the build.sh files for missing dependencies and configurations, a  lot of stuff might already be in place since calibre-web is alread installed
- try to figure out the correct locations inside the lxc container for these new or updates files
- apply/copy over missing dependencies and  configurations from the git repo files
- copy over the new Pyhton files to the appropriate locations
- give it a try and see if it starts....

There are a few things inside these docker files I do not understand and how to port them:
What is /etc/s6-overlay for? is this docker specific? can this be done inside a regular debian 12 lxc container aswell?

If all this manual procedure works it might be another option to update the regular lxc calibre-web installer with these steps later on.

I would love to do an updated lxc installer for calibre-web-automated, but there are a few things to learn first about the structures of those files in docker and LXC I am afraid.

What do you guys think? Does it make sense?

**Comment by AJErazzor** [comment]
following

**Comment by anaxmedia** [comment]
+1 for Calibre Web Automated would love to have a script for this!

**Comment by vhsdream** [comment]
This is going to be more work than you think. I've managed to cobble together something that appears to work on the surface (here is a screen of the CWA settings page):
![image](https://github.com/user-attachments/assets/3cb0c24c-ccf6-4b3b-a0a6-c33efc2a4f76)

But in order to even get this far, I had to fork the CWA repo and modify a number of file paths in several files. I'm not sure of a way to have a helper-script (or a function included in the existing Calibre-Web helper-script) modify and keep track of so many changes in so many files, especially when CWA is still being developed. There are so many moving parts that things are bound to break.

If you're willing to take this one further (I think I'll keep working on it, but my skill is very limited), then you can fork my dev branch [here](https://github.com/vhsdream/cwa-lxc/tree/dev). The install script (cwa_install.sh) is a WIP and although the screenshot above makes it look like things are working - not all (or none) of the services that make CWA CWA are working yet.

**Reply by vhsdream** [reply]
### Update: install script can be found in [this gist](https://gist.github.com/vhsdream/58fb87c4b10e8fd5a09b3cfcf99cb4f0).
### Use at your own risk.
### Don't run this on your 'production' Calibre-Web install.
### Downloads a modified fork of CWA from [my dev branch](https://github.com/vhsdream/cwa-lxc/tree/dev).

Here is what I have found is working (or appears to be working) so far:
- Auto-ingest
- Auto-enforcement of covers & metadata
- Light mode/dark mode switching
- Automated backup service (I think?)
- EPUB Fixer service
- Library Autodetect
- Stats page
- Manual library refresh (I think?)

I have not tested the below features to see if they are also working:
- Enabling/disabling CWA features via the UI
- Update notifications (and probably updates)
- Nightly automatic backups with compression
- Bulk editing
- Probably other things

As is, I don't think this is something the Proxmox Helper Script folks should condone or support since it relies on a customized fork that I have no intention of maintaining.

A more robust, secure and simple way of modifying files should be used, and I am hopeful we can all come up with a solution together. Perhaps through a frankly ridiculous amount of `sed` commands?

**Reply by vhsdream** [reply]
### Update: Install script [updated](https://gist.github.com/vhsdream/58fb87c4b10e8fd5a09b3cfcf99cb4f0)

The script in the gist has been updated a bit to clean some things up; but the biggest change is that rather than downloading from my fork, it uses a git patch file to modify the necessary files, after cloning the Calibre-Web Automated repo and switching to the latest release. The patch file is still downloaded from my repo though, but using a patch might be the best way to deal with this install.

Is hosting and managing a patch file something the Community-Script devs would be willing to do, or is it outside the purview of this repo? I've also asked the maintainer of the CWA repo if they want to contribute.

**Comment by btehan** [comment]
https://github.com/vhsdream/cwa-lxc?tab=readme-ov-file#proxmox-users-migrating-from-a-calibre-web-lxc-installation

For step i: 

In the calibre-web LXC console:
wget https://gist.githubusercontent.com/vhsdream/58fb87c4b10e8fd5a09b3cfcf99cb4f0/raw/ee1cef59eed59cf6d83c100fe6b116d95c10d8ba/cwa_install.sh


Thanks!

**Comment by vhsdream** [comment]
Yeah I wouldn't use that anymore, it's unfinished and stil has some bugs.

The one that is complete and will get you a fully working CWA LXC (not simply convert Calibre-Web, this time from scratch) is [here](https://github.com/community-scripts/ProxmoxVED/blob/cwa-func/ct/calibre-web-automated.sh), but it is not going to be released until we find a better way of implementing the numerous changes to the file paths etc etc. It relies on a func file that the maintainers do not want added to the repo.

The func file is [here](https://github.com/community-scripts/ProxmoxVED/blob/cwa-func/misc/cwa_patcher.func) so you can see what it does to the Calibre-Web Automated Docker version - always a good idea to know what the scripts are doing 😄

**Reply by Tooelite** [reply]
Can you get in detail or do you know the details what needs to be improved so the main devs will accept the way files are patched?

**Reply by tremor021** [reply]
We know what he needs to add/change, but the problem is that those changes are in core parts of our script backend. We think that no script needs to change the backend in order to be installed/supported. Actually this is the only script that actually requires such changes.

When we find a good, reliable way of deploying CWA, without messing with core of our scripts, we will happily support it.

**Reply by vhsdream** [reply]
Yep, it's basically that there are so many operations that need to be done "post-install" that it would end up making the install script (and the update function) really unwieldy if the operations were to be moved into them and not into a separate func file like I proposed; not to mention the fact that when CWA Docker gets updated it is likely someone would need to go through the code to see what else the script needs to change. With hundreds of helper scripts to maintain, it's understandable that it is not approved in it's current state.

So it should go without saying that if you're gonna use the script in the branch above, it is at your own risk etc etc etc.

**Comment by vhsdream** [comment]
So I created a [repo](https://github.com/vhsdream/calibre-web-automated-lxc) for my version of the script that you can run in an existing Debian 12 LXC. It does not upgrade an existing Calibre-Web LXC and it does not update; the latter will be added later, and the former I'm not sure about.

It borrows heavily from what I've learned from this repo since I started contributing, so it's not ugly and it works 😸 
![image](https://github.com/user-attachments/assets/09d71d87-f2f3-484a-bd0b-0275ac0e2751)

**Reply by Tooelite** [reply]
Awesome! Thank you!
